Compute recent lightclient updates

dapplion opened the pull request:
Current lightclient implementation is reactive, it produces updates in response to external queries.

This PR allows the beacon node to eagerly produce lightclient updates immediately after processing each block. This approach reduces the overall cost to serve lightclient data since each update is computed at most once. Lightclient updates need access to full states to compute proofs, by computing the updates immediately after block production block replay is never necessary in normal network conditions.

In summary lightclient updates are produced in two steps:
1. cache state proofs from the attested block
2. use the next block's sync aggregate to produce the update

![image](https://github.com/sigp/lighthouse/assets/35266934/bb4741e8-c0ae-4f7d-a839-b6ca90bbbf24)

to not delay the import_block routine, the second step in done in a separate thread with access to the network senders to push gossip updates in a latter PR.

Since this PR only produces recent updates, it's not necessary to persist them to disk. Gossip validation rules are simplified significantly since now you just need an equality check with the most recently produced update.

## Issue Addressed

- https://github.com/sigp/lighthouse/issues/4926

Closes https://github.com/sigp/lighthouse/issues/4926

## Next steps

- Wire recent updates producer to gossip and SSE https://github.com/dapplion/lighthouse/compare/lc-prod-recent-updates...dapplion:lighthouse:lc-prod-push-recent-updates?expand=1

Commit:
Compute recent lightclient updates

jimmygchen commented on beacon_node/client/src/broadcast_lightclient_updates.rs:
```diff
@@ -0,0 +1,30 @@
+use slog::{error, Logger};
+use tokio::sync::mpsc::Receiver;
+
+// Each LightclientProducerEvent is ~200 bytes. With the lightclient server producing only recent
+// updates it is okay to drop some events in case of overloading. In normal network conditions
+// there's one event emitted per block at most every 12 seconds, while consuming the event should
+// take a few miliseconds. 32 is a small enough arbitrary number.
```
small typo:

```suggestion
// Each `LightclientProducerEvent` is ~200 bytes. With the lightclient server producing only recent
// updates it is okay to drop some events in case of overloading. In normal network conditions
// there's one event emitted per block at most every 12 seconds, while consuming the event should
// take a few milliseconds. 32 is a small enough arbitrary number.
```

jimmygchen commented on beacon_node/client/src/builder.rs:
```diff
@@ -199,6 +206,16 @@ where
         };
 
+        let builder = if config.network.enable_light_client_server {
+            let (tx, rv) = tokio::sync::mpsc::channel::<LightclientProducerEvent<TEthSpec>>(
```
I don't know if there's been any discussion on whether we use `tokio::sync::mpsc` vs `futures::channel::mpsc`, and it seems like both are currently used in Lighthouse. I don't know if we prefer one over another, It might be worth considering sticking with `futures::channel::mpsc` for now since it's the one used in `beacon_chain` to avoid using both types in the same file? (cc @michaelsproul)

jimmygchen replied:
Just reading the docs and it looks like the `sender.try_send` for both crates behaves similarly - returns immediately and will return error if the buffer is full, which is what we want I think.

dapplion replied:
Was debating both options at the time, happy to go with futures for consistency

jimmygchen commented on beacon_node/client/src/broadcast_lightclient_updates.rs:
```diff
@@ -0,0 +1,30 @@
+    mut lightclient_server_rv: Receiver<LightclientProducerEvent<T::EthSpec>>,
+    log: &Logger,
+) {
+    // lightclient_server_rv is Some if lightclient flag is enabled
+    //
```
I think this comment is no longer relevant as rv is not an `Option` here

```suggestion
```

jimmygchen commented on beacon_node/client/src/broadcast_lightclient_updates.rs:
```diff
@@ -0,0 +1,30 @@
+    // Intents to process SyncAggregates of all recent blocks sequentially, without skipping.
+    // Uses a bounded receiver, so may drop some SyncAggregates if very overloaded. This is okay
+    // since only the most recent updates have value.
+    while let Some((block_root, slot, sync_aggregate)) = lightclient_server_rv.recv().await {
+        chain
+            .recompute_and_cache_lightclient_updates((block_root, slot, sync_aggregate))
```
nit: looks like we don't need to destructure the event here:

```suggestion
    while let Some(event) = lightclient_server_rv.recv().await {
        chain
            .recompute_and_cache_lightclient_updates(event)
```

jimmygchen commented on beacon_node/beacon_chain/src/lightclient_proofs_cache.rs:
```diff
@@ -0,0 +1,247 @@
+/// requests. These messages include proofs on historical states, so on-demand computation is
+/// expensive.
+///
+pub struct LightclientServerCache<T: BeaconChainTypes> {
```
The file can be renamed to `lightclient_server_cache.rs`

jimmygchen commented on beacon_node/beacon_chain/src/light_client_optimistic_update_verification.rs:
```diff
@@ -110,30 +85,21 @@ impl<T: BeaconChainTypes> VerifiedLightClientOptimisticUpdate<T> {
-            parent_root: canonical_root,
+            light_client_optimistic_update: rcv_optimistic_update,
+            // TODO: why is the parent_root necessary here?
+            parent_root,
```
Looks like the `Ok` output of this function is only used for logging below, and we don't use any of these output vallues, so we could potentially refactor this to just return `Result<(), Error>`

https://github.com/sigp/lighthouse/blob/ba891e1fed8379f1410f50ce56eccc05f25daac2/beacon_node/network/src/network_beacon_processor/gossip_methods.rs#L1751-L1759

dapplion replied:
Sure, let's do that in a future pr dealing with gossip

jimmygchen commented:
Hey @dapplion thanks for this PR. I haven't got a chance to look at the gossip verification changes yet - will continue tomorrow, but overall looks great so far!

jimmygchen commented on beacon_node/beacon_chain/src/builder.rs:
```diff
@@ -912,6 +924,8 @@ where
             validator_pubkey_cache: TimeoutRwLock::new(validator_pubkey_cache),
             attester_cache: <_>::default(),
             early_attester_cache: <_>::default(),
+            lightclient_server_cache: LightclientServerCache::new(),
```
Can we make this an `Option`  so we only compute and cache the proofs if light client server is enabled?

dapplion replied:
I agree that conceptually looks cleaner however adding an option requires to handle the case where HTTP API and ReqResp handlers receive a request but the lightclient server in disabled. We can achieve the same purpose by only adding data into the lightclient_server_cache if a config value is true. In the case the lightclient server is disabled there will be an extra cache in memory, but I think overall it reduces complexity. As long as we make sure that the network flag disabling the lightclient routes is the also used in the BeaconChain config. No strong opinion either way tho

jimmygchen replied:
Yeah cool, I'm happy either way as long as we don't perform the computation if light client server is disabled.

jimmygchen commented on beacon_node/beacon_chain/src/lightclient_proofs_cache.rs:
```diff
@@ -0,0 +1,247 @@
+    /// Tracks a single global latest finality update out of all imported blocks.
+    ///
+    /// TODO: Active discussion with @etan-status if this cache should be fork aware to return
+    /// latest canonical instead of global latest.
```
What do you mean by global latest vs latest canonical here?

dapplion replied:
Latest canonical: update with highest signature slot, where its attested header is part of the head chain
Global latest: update with highest signature slot, out of all branches

jimmygchen commented on beacon_node/client/src/broadcast_lightclient_updates.rs:
This file should probably be renamed to `compute_lightclient_updates.rs`

jimmygchen commented on beacon_node/beacon_chain/src/light_client_finality_update_verification.rs:
```diff
@@ -110,24 +94,13 @@ impl<T: BeaconChainTypes> VerifiedLightClientFinalityUpdate<T> {
         // verify that the gossiped finality update is the same as the locally constructed one.
-        if finality_update != light_client_finality_update {
+        if latest_finality_update != rcv_finality_update {
             return Err(Error::InvalidLightClientFinalityUpdate);
```
Not part of this PR, but this validation error is treated as a low tolerance error (peer gets downscored significantly) - does it make sense to just downscore slightly on this one?

https://github.com/sigp/lighthouse/blob/ba891e1fed8379f1410f50ce56eccc05f25daac2/beacon_node/network/src/network_beacon_processor/gossip_methods.rs#L1690-L1703

dapplion replied:
Yes agree, clients can naturally disagree if one process the head update latter than another. It's not a very adversarial action

jimmygchen commented on beacon_node/beacon_chain/src/light_client_optimistic_update_verification.rs:
```diff
@@ -67,38 +65,15 @@ impl<T: BeaconChainTypes> VerifiedLightClientOptimisticUpdate<T> {
-            None => Slot::new(0),
-        };
-
-        // verify that no other optimistic_update with a lower or equal
-        // optimistic_header.slot was already forwarded on the network
-        if gossiped_optimistic_slot <= latest_seen_optimistic_update_slot {
-            return Err(Error::OptimisticUpdateAlreadySeen);
-        }
```
Why is this check no longer required?

dapplion replied:
This check is redundant with the equality done below, I'll also removed it from the finality update verif fn

https://github.com/sigp/lighthouse/blob/90cf69e98fe0ec1765418aad3b4a09c8dd8d744a/beacon_node/beacon_chain/src/light_client_optimistic_update_verification.rs#L94-L96

jimmygchen replied:
Nice, we can probably get rid of the `Error` enum variants too
https://github.com/sigp/lighthouse/blob/a5ba62da2caa789941ed4354c35449eb7fef2efb/beacon_node/beacon_chain/src/light_client_finality_update_verification.rs#L17-L18
and 
https://github.com/sigp/lighthouse/blob/a5ba62da2caa789941ed4354c35449eb7fef2efb/beacon_node/beacon_chain/src/light_client_optimistic_update_verification.rs#L18-L19

dapplion committed:
Review PR

dapplion committed:
Merge remote-tracking branch 'upstream/unstable' into lc-prod-recent-updates

jimmygchen commented on beacon_node/beacon_chain/src/lightclient_server_cache.rs:
```diff
@@ -0,0 +1,249 @@
+    }
+
+    /// Compute and cache state proofs for latter production of light-client messages. Does not
+    /// trigger block replay. May result in multiple DB write ops.
+    /// TODO: Should return StoreOps to batch with rest of db operations?
```
We're not writing to DB here, is this the plan for future PRs?

jimmygchen commented on beacon_node/src/config.rs:
```diff
@@ -156,6 +156,10 @@ pub fn get_config<E: EthSpec>(
     }
 
+    if cli_args.is_present("light-client-server") {
+        client_config.chain.enable_lightclient_server = true;
```
Please add a `CommandLineTest` to cover this, similar to:
https://github.com/sigp/lighthouse/blob/c8b2324880ef3d3dd5f9233285cb59fc3701db71/lighthouse/tests/beacon_node.rs#L2364-L2374

dapplion replied:
Added

https://github.com/sigp/lighthouse/blob/6dc45c4d8883d70f7f83c7f1601cdd7fa348a33a/lighthouse/tests/beacon_node.rs#L2355-L2364

jimmygchen commented:
Nice updates, thanks ☺️  I've left a few more questions & comments, but I think we're close!

I've got a few more things in mind, but happy to address these in a separate PRS:
- It would be useful to capture some metrics on computing light client data, this will help evaluate the readiness to make light client data serving the default behaviour.
- Nit: we've unified existing code to separate the "light client" into two words (i.e. `light_client` or `LightClient`), we should keep the consistency
- We don't have any tests rn, we should add a test on the `beacon_chain` to verify the light client data computation.

jimmygchen commented on beacon_node/beacon_chain/src/lib.rs:
```diff
@@ -58,9 +59,8 @@ pub mod validator_pubkey_cache;
     BeaconBlockResponseWrapper, BeaconChain, BeaconChainTypes, BeaconStore, ChainSegmentResult,
-    ForkChoiceError, OverrideForkchoiceUpdate, ProduceBlockVerification, StateSkipConfig,
-    WhenSlotSkipped, INVALID_FINALIZED_MERGE_TRANSITION_BLOCK_SHUTDOWN_REASON,
-    INVALID_JUSTIFIED_PAYLOAD_SHUTDOWN_REASON,
```
Looks like this was removed unintentionally, hence CI job failing here
https://github.com/sigp/lighthouse/actions/runs/7170965412/job/19538151833?pr=4969

dapplion committed:
Review PR

dapplion committed:
consistent naming

dapplion commented:
> It would be useful to capture some metrics on computing light client data, this will help evaluate the readiness to make light client data serving the default behaviour.

Right, I miss some metrics in this PR too. Should not be too many extra lines so I'll add it.

EDIT: Added three metrics:
- `LIGHT_CLIENT_SERVER_CACHE_STATE_DATA_TIMES`: track that cache data fn is actually called and how expensive it is computationally
- `LIGHT_CLIENT_SERVER_CACHE_RECOMPUTE_UPDATES_TIMES`: track the produce event is fired, and events recomputed + how expensive it is
- `LIGHT_CLIENT_SERVER_CACHE_PREV_BLOCK_CACHE_MISS`: tracks how many calls into recompute_and_cache_updates() result in a cache miss that requires recreating a state.

Successful metrics values would be very short / negligible times for the two histograms and zero cache misses

> Nit: we've unified existing code to separate the "light client" into two words (i.e. light_client or LightClient), we should keep the consistency

Renamed all occurrences into `light_client`, `LightClient`, `LIGHT_CLIENT`

> We don't have any tests rn, we should add a test on the beacon_chain to verify the light client data computation.

Where should we place such test? Could be something like

```
chain.import_block(block)
let update = chain.light_client_server_cache.get_latest_optimistic_update()
assert update == update_from_block(block)
```

dapplion committed:
add metrics

jimmygchen commented:
> Where should we place such test? Could be something like

Great updates! I think we can add the test to a new file dedicated for light clients under `beacon_node/beacon_chain/tests`. It should be pretty easy to achieve with `BeaconChainHarness`,  we could use `harness.extend_chain` with some blocks and access the beacon chain using `harness.chain. light_client_server_cache`. This would be useful as well when we make the light client server fork-aware.

jimmygchen commented on beacon_node/beacon_chain/src/light_client_optimistic_update_verification.rs:
```diff
@@ -67,73 +43,36 @@ impl<T: BeaconChainTypes> VerifiedLightClientOptimisticUpdate<T> {
+            < start_time + one_third_slot_duration
+        {
+            return Err(Error::TooEarly);
         }
```
I had a brief discussion with @michaelsproul regarding the rationale behind re-processing the light client updates (added in https://github.com/sigp/lighthouse/pull/3799). 

The current logic is that if we receive the light client update before receiving the new block, this would result in the message being queued for re-processing later. The point is that we would still validate the gossip message later so it propagates, rather than telling gossip to IGNORE the message straight up, which might hinder their propagation.

With the new logic here, beacon node would compute the update after block import, and we would publish this on gossip, but the `IGNORE` validation result propagated earlier could still prevent this from being propgated?

It seems like we might want to keep this `UnknownBlockParentRoot` check and re-processing logic differentiate this with `InvalidLightClientOptimisticUpdate`.

dapplion replied:
Note: the re-processing logic should then apply to both optimistic update and finality updates, both will be broadcast when signature slot advances https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/full-node.md#create_light_client_optimistic_update

dapplion replied:
Reverted dropping the reprocessing logic in https://github.com/sigp/lighthouse/pull/4969/commits/b41ed0759964990bec486e223cc3a25ae2c93377

However the re-processing logic should be changed. In unstable when the `BlockImported` event is received by the reprocessing queue the validation logic can then create the latest update and pass validation. However with this PR only when the `recompute_and_cache_light_client_updates` fn completes, the validation logic will pass.

dapplion replied:
The optimistic message validation rules require an equality on the entire message, including the sync aggregate which is dependent on the block at the signature slot, not its parent. Since the validity of the message depends on our local view of the current head it's not useful to queue by branch (parent_root).

Consider a case with heavy forking where there are two valid proposers at slot N in their respective branches. If you receive two optimistic updates with the same signature slot and different attested headers you will only propagate the one that matches your local view. Similarly, a missed slot means there's no possible valid optimistic update.

So the reprocess queue for optimistic updates at slot N should wait to complete running `recompute_and_cache_light_client_updates` on _any_ branch for slot N and expire after 1 slot.

jimmygchen replied:
Got it, so it looks like we need to rework most of the re-reprocessing logic for light client updates.

Perhaps we can do this:
1. Add a new `ReprocessQueueMessage` called `NewLightClientUpdate`
2. Send this event to reprocessing queue after we compute a new update (upon completion of `recompute_and_cache_light_client_updates`)
3. Move the existing event handling logic from [`BlockImported` match arm](https://github.com/sigp/lighthouse/blob/33976121601d9d31d26c1bb2e1f0676e9c4d24fc/beacon_node/beacon_processor/src/work_reprocessing_queue.rs#L691-L739) to a new match arm for  `InboundEvent::Msg(NewLightClientUpdate)`. This marked it as ready for re-processing
4. A re-processed LC update will then either get _ACCEPTED_ or _IGNORED_.

Any `QueuedLightClientUpdate` that is not reprocessed is expired after 12 seconds ([delay queue](https://github.com/sigp/lighthouse/blob/33976121601d9d31d26c1bb2e1f0676e9c4d24fc/beacon_node/beacon_processor/src/work_reprocessing_queue.rs#L611-L614))

jimmygchen replied:
> Since the validity of the message depends on our local view of the current head it's not useful to queue by branch (parent_root).

~~Would it make more sense to queue it by `block_root` or`block_slot`? It feels like the selection of either wouldn't matter too much since anything that's not reprocessed within a slot is expired. Not sure about the re-org case though~~

We don't really have the head block root in the optimistic update, so `parent_root` would be the best thing we could use to key the message for the queue I think.

jimmygchen replied:
I made an attempt to update the reprocessing logic based on the above: https://github.com/dapplion/lighthouse/pull/7

I've just came across @michaelsproul's [comment here](https://github.com/sigp/lighthouse/pull/3799#discussion_r1084699605) which might be relevant to the parent root check here. I don't fully understand it, but it's too late here now so I'll revisit tomorrow.

dapplion replied:
Suggestion looks great, merged into this branch

dapplion committed:
revert dropping reprocessing queue

jimmygchen committed:
Update light client optimistic update re-processing logic. (#7)

dapplion commented:
> > Where should we place such test? Could be something like
> 
> Great updates! I think we can add the test to a new file dedicated for light clients under `beacon_node/beacon_chain/tests`. It should be pretty easy to achieve with `BeaconChainHarness`, we could use `harness.extend_chain` with some blocks and access the beacon chain using `harness.chain. light_client_server_cache`. This would be useful as well when we make the light client server fork-aware.

Adding a test in `beacon_node/beacon_chain/tests` won't be complete as the harness does not start the extra thread that calls `compute_light_client_updates` in the beacon_node builder.

https://github.com/sigp/lighthouse/blob/31acd0c2cf84ab2ab95a1006f3d9eb6300af45b2/beacon_node/client/src/builder.rs#L860

The chain instance inside the harness has None as light_client_server_tx so no update is ever produced. Is there a way to recreate the more "complete" beacon node setup without repeating code in the harness?

jimmygchen committed:
Add light client server simulator tests. Co-authored by @dapplion.

jimmygchen committed:
Merge branch 'unstable' into fork/dapplion/lc-prod-recent-updates

jimmygchen committed:
Fix lint

jimmygchen committed:
Enable light client server in simulator test.

jimmygchen committed:
Fix test for light client optimistic updates and finality updates.

jimmygchen commented:
@dapplion and I got a passing test in the simulator, which shows that it's constantly producing optimistic updates and finality updates as the chain progresses! 🎉 

I haven't looked at the metrics yet on how efficient the caching is, but I think we'll be able to start running some tests on a testnet once we have #4946 (Capella & Deneb light client support) merged as well.

Ready for a final review!

jimmygchen reviewed (approved):
I've re-reviewed the PR again - I think there's no changes to the regular code path when `ligth-client-server` is disabled, and we've got altair light client updates available according to the simulator tests,  so I'm happy with the changes now 👍  thanks @dapplion!

dapplion reviewed (commented):
Can't approve my own PR, but the last string of commits from @jimmygchen look great https://github.com/sigp/lighthouse/pull/4969/files/b41ed0759964990bec486e223cc3a25ae2c93377..a63a4886d140d19506c73ee54c4f6971654db9ad

thanks for nailing the tests!

jimmygchen commented:
@mergifyio queue

michaelsproul commented:
@mergifyio requeue
